web/src/typing.ts

```diff
@@ -47,16 +55,16 @@ function send_stream_typing_notification(stream_id, topic, operation) {
     send_typing_notification_ajax(data);
 }
 
-function send_typing_notification_based_on_message_type(to, operation) {
-    const message_type = to.stream_id ? "stream" : "direct";
+function send_typing_notification_based_on_message_type(to: Recipient, operation: string): void {
+    const message_type = "stream_id" in to ? "stream" : "direct";
     if (message_type === "direct" && user_settings.send_private_typing_notifications) {
         send_direct_message_typing_notification(to, operation);
-    } else if (message_type === "stream" && user_settings.send_stream_typing_notifications) {
+    } else if ("stream_id" in to && user_settings.send_stream_typing_notifications) {
```

Comment:
TypeScript can not figure out whether `to` contains `stream_id` property by mere `message_type==="stream"` check.

Reply:
Let's keep the check readable as originally written -- checking the message type.

I think a `!` type assertion or an explicit assert would be better for informing TypeScript that this means there's a `stream_id` field in the object ... though probably even better would be to improve the typing notifications types so it knows already.

Reply:
Hmm I guess the best fix would be to do a prep commit extending the `Recipient` type in `typing_status.ts` to always be an object with an explicit `message_type` field, and be a union of two types, one with `message_type=stream` and the other with `message_type=direct`.

Reply:
Extending `Recipient` to always be an object effectively means to re-type `typing_status.ts` as a lot of function leverage the fact of `Recipient` being` number[]` type sometimes. To maintain check readability  I think we can create a new type with disciminated union with structure same as of your proposed `Recipeint` type ie. an object with an explicit `message_type` property ……

Reply:
I got no problem with mutating `Recipient` but I think it would be good if we can come up with something else for a problem which is local to this function block only.
Would go with your thoughts ......

Reply:
I agree that extending `Recipient` type is a good refactor in general but `typing_status.ts` is also being used the mobile app. The changes on the web app won't be complex but I don't remember the details on how this would work for the mobile app. If it's a lot of hassle co-ordinating this with the mobile app, I'd lean towards fixing it here itself

Reply:
Maybe it makes sense for `StreamTopic` to be a class so that we use `to instanceof StreamTopic` and keep this function clean?